OpenBSD support

mprins opened the pull request:
OpenBSD support


resolve #1465

dbwiddis commented:
I went ahead and did a mass copy of all the other files.  Let me know what doesn't work and I can help you sort it out.

dbwiddis commented:
OK, so I did manage to get a minimal openbsd VM set up and see what you mean with JNA -- I can't for the life of me figure out how to load the C library.

dbwiddis commented:
OK, at this point SystemInfoTest runs without crashing.  All those nulls and empty arrays you added were causing headaches. :) 

I will call that a win for now. Let me know which data is most important for you.

dbwiddis commented:
OK, I've got the C library loading. It was just confusion trying to open the "FreeBSD" library with poorly linked structures.  OpenBSD does have `sysctl()` but not `sysctlbyname()` which means we can make most of the native calls we need, we just need to fetch the constants.  Dropping the link here for later reference:  http://fxr.watson.org/fxr/source/sys/sysctl.h?v=OPENBSD

mprins commented:
Cool!
Away from computer atm but https://github.com/openbsd/src/blob/master/sys/sys/sysctl.h is the current version

Will give it a shot in the morning

dbwiddis commented:
Current status: 
 - Native sysctl should work for most KERN and HW queries. Not sure what MIB values are for the MACHDEP and VM queries yet as they are not in the header file.
 - Using native sysctl for CPU ticks and frequency.  For some reason `hw.ncpu` fails in the native version so I fell back to your commandline workaround there.
 - Parsed the CPUID to get family, model, stepping.
 - CentralProcessor looks done except for context switches and interrupts.

Next up is memory, available from sysctl `CTL_VM` , `VM_UVMEXP`, returning a uvmexp structure, which looks like it's just a pointer and manual offsets are needed. See https://github.com/openbsd/src/blob/master/usr.bin/systat/uvm.c and https://github.com/openbsd/src/blob/master/usr.bin/vmstat/vmstat.c#L321 for all sorts of porting fun.

dbwiddis commented:
I think I'm mostly done with processor and memory. Next big thing to tackle is the process listing.   There may be a way to do it natively, but I keep running into problems with those (possibly due to OpenBSD's famed security measures) so parsing ps is probably the way to go.  The existing code produces *something* but it's clearly out of sync on the columns, so you might take a look at where the FreeBSD and OpenBSD commands differ and tweak the process (and possibly thread) code.

dbwiddis commented:
To Do list:
- [ ] Test if you have elevated permissions. I think `id -u` works here, steal code from AixOperatingSystem.
- [ ] Processes & threads: review where ps parsing is broken and fix
- [ ] PowerSupply: `apm` command or sysctl `hw.sensors.acpisbs0.*`
- [ ] Disks & filesystem: `mount`, `df`, etc.?
- [ ] IP stats: parse `netstat -s` output
- [ ] Displays
- [ ] USB Devices
- [ ] Sound and Graphics cards

mprins commented:
I think I've figured out most of the process things, not the threads stuff yet (not micht is exposed through `ps`)
also fixed up elevated priv. call and fixed cpu frequency reporting.

Not sure what to report for number of CPU's (`hw.ncpuonline` or `hw.ncpu` / `hw.ncpufound`) OpenBSD has HT/SMT turned off by default (`hw.smt=0`), so basically 2 logical processors turned off. Eg. on my laptop I have:
```
mprins @ thinkpad ~ 
└─ $ ▶ sysctl hw
hw.machine=amd64
hw.model=Intel(R) Core(TM) i7-4600M CPU @ 2.90GHz
hw.ncpu=4
hw.byteorder=1234
hw.pagesize=4096
hw.disknames=sd0:2cf69345d371cd82,cd0:
hw.diskcount=2
hw.sensors.cpu0.temp0=52.00 degC
hw.sensors.acpibtn0.indicator0=On (lid open)
hw.sensors.acpiac0.indicator0=On (power supply)
hw.sensors.acpithinkpad0.temp0=54.00 degC
hw.sensors.acpithinkpad0.temp1=54.00 degC
hw.sensors.acpithinkpad0.temp2=54.00 degC
hw.sensors.acpithinkpad0.temp3=54.00 degC
hw.sensors.acpithinkpad0.temp4=54.00 degC
hw.sensors.acpithinkpad0.temp5=54.00 degC
hw.sensors.acpithinkpad0.temp6=54.00 degC
hw.sensors.acpithinkpad0.temp7=54.00 degC
hw.sensors.acpithinkpad0.fan0=0 RPM
hw.sensors.acpithinkpad0.indicator0=Off (port replicator), UNKNOWN
hw.sensors.acpitz0.temp0=54.00 degC (zone temperature)
hw.cpuspeed=800
hw.setperf=0
hw.vendor=LENOVO
hw.product=20AWA08J00
hw.version=ThinkPad T440p
hw.serialno=PB011B2U
hw.uuid=81ea2b95-f453-cb11-afbc-b7c3250e8200
hw.physmem=8446279680
hw.usermem=8266711040
hw.ncpufound=4
hw.allowpowerdown=1
hw.perfpolicy=auto
hw.smt=0
hw.ncpuonline=2

```
so only 2 real cpu cores (and eg. `top` reports those as CPU0 and CPU1, but as you can see 4 are "detected", see also my commit e6b833639b7594effcf183dda04f2ef9a9ad9215 which now reports cpu freq as `{N,0,N,0}`

dbwiddis commented:
> Not sure what to report for number of CPU's (`hw.ncpuonline` or `hw.ncpu` / `hw.ncpufound`) OpenBSD has HT/SMT turned off by default (`hw.smt=0`), so basically 2 logical processors turned off. 

This is one of those fuzzy situations where we have a real physical reality (the CPUs on the system) and then "software" that alters it.  If I were re-designing OSHI, I'd probably try to separate those concepts, but at this point I think we're stuck.

Given that we assume the processor count never changes during operation (unless you start with a new `SystemInfo` object) I think we have to keep the total "found" number.  It's interesting that the frequency (and probably per-CPU tick count) goes on the 0 and 2 indices, skipping 1 and 3, indicating "you really do have 4 logical processors, you're just not using two of them".  

However, in that interpretation it's still "2 physical processors" and "4 logical processors".  Currently I just hardcoded "1" into the physical count value, but it should probably be hardcoded as "logical/2" given this discussion, or, detect the state of hyperthreading?  Or use "ncpufound" if hw.smt is 0, and use "ncpufound/2" if hw.smt is 1?

dbwiddis commented:
For threads, looks like the columns displayed on `ps -axH` are available for per-process, e.g., `ps -axHo pid,tid,tt,stat,time,command` which can be reasonably useful for the thread details.

dbwiddis commented:
For power supply, [this blog post](http://karlbsd.blogspot.com/2010/01/openbsd-nc4200-howto-get-battery-status.html) from 10 years ago lists `hw.sensors.acpibat0.*` as the source but my VM has `hw.sensors.acpisbs0.*` for the same info.  Should we check both? I don't see anything listed for your laptop.

dbwiddis commented:
I'm going to take a crack at updating the Sensors and Power Supply classes, using `systat -d1 sensors` to grab info from the appropriate sysctls for the various drivers.  And I'll update the IP stats from a `netstat -s` call.

dbwiddis commented:
OK, done with sensors, power source, and ip statistics.  Lots of "missing" stuff for the ip stats that I couldn't find a correlating entry, and one that's not a constant string so I cheated (for now) and left it blank.  I think I've gone about as far as I can go with my VM... left some notes in the TODO on ways to approach disks and partitions.   No idea about USB, display (unless `xrandr` works), and audio/graphics cards.  So basically this PR is all yours now... let me know if you need more help with anything.

dbwiddis commented:
Tweaked missing stuff in memory and IP stats.   Also made an attempt to do the USB device listing based entirely on reading the source code here: https://github.com/openbsd/src/blob/master/usr.sbin/usbdevs/usbdevs.c

I don't have any devices with which to test it, though, so you'll need to test and (if necessary) debug that.

dbwiddis commented:
For USB devices, the format I parsed is the current one.  Previous format [looks like this](https://unix.stackexchange.com/questions/71969/webcam-on-openbsd).... I'll work tomorrow to be compatible with the older format as well.

Here's [the diff](https://github.com/openbsd/src/commit/d412bb608ace9441451c28fdbeb680133ebf9aee#diff-236278e4c176081169c22bd2c64791ec3791e88d569db0bf41151a6ac0e97f08) of the format change.

mprins commented:
I wouldn't worry about "old" stuff; OpenBSD has 2 releases per year, anything older than  a year+ is considered unsupported in the OpenBSD community

dbwiddis commented:
> I wouldn't worry about "old" stuff

Fair. I think that went in in 6.4, so I think we can just state that's our minimum level.  (I did do pre-6.4 for CPU usage...)

As an alternative which happens to be compatible across all versions it may be that a native `ioctl()` implementation may work, so I may try that too (as the structure would be easier than string parsing).

dbwiddis commented:
If you have an actual listing with the current format that you could paste here it would help me test.

dbwiddis commented:
Updated To Do list:
- [x] Test if you have elevated permissions. I think `id -u` works here, steal code from AixOperatingSystem.
- [ ] Processes & threads: review where ps parsing is broken and fix
  - this may be done?  One outstanding question regarding using `command` for thread names
- [x] Sensors: `systat -ab sensors`
- [x] PowerSupply: `apm` command or sysctl `hw.sensors.**` via `systat -ab sensors`.
- [ ] Disks & filesystems: `sysctl hw.disknames`, `disklabel -h <diskname>` (root permissions), also gives partitions, non-root parse `dmesg` for lines starting `<diskname> at ...` for more info, `df -h` lists mounted partitions/filesystems and correlates to disks. See https://www.romanzolotarev.com/openbsd/disk.html and https://www.cyberciti.biz/faq/openbsd-mounting-usb-flash-drive-harddisk/
- [x] IP stats: parse `netstat -s` output
- [ ] Displays: may just use existing `xrandr` based
- [x] USB Devices: this is done based on reading source code for `usbdevs -v` but needs testing  
- [ ] Sound and Graphics cards: 
  - try parsing `dmesg` and looking for devices with "audio* at" etc. 
  - Possibly `audioctl -f /dev/audio0` works (list all `/dev` directory contents with `audio*`) for something really basic
  - Also `/dev/sound*` ...

dbwiddis commented:
Threw together something for video and sound cards.  Also while digging into `dmesg` saw where it correlated logical processors with cores and packages.  Of note, it appears the non-smt cores are listed first and smt are listed second.  See [this pastebin](https://pastebin.com/T3UGucB1):
```
cpu0: smt 0, core 0, package 0
cpu1: smt 0, core 1, package 0
cpu2: smt 1, core 0, package 0
cpu3: smt 1, core 1, package 0
```
So I'd suggest for your CPU frequencies you put them at index 0 and 1 rather than index 0 and 2.

Only disk and filesystems left... still trying to leave that for you since you have actual hardware.

mprins commented:
family things going on tis weekend; I'll look into disk/filesystem thing later during the week.

here's what `usbdevs -v` says:
```
└─ $ ▶ usbdevs -v
Controller /dev/usb0:
addr 01: 8086:0000 Intel, EHCI root hub
	 high speed, self powered, config 1, rev 1.00
	 driver: uhub0
addr 02: 8087:8008 Intel, Rate Matching Hub
	 high speed, self powered, config 1, rev 0.04
	 driver: uhub2
addr 03: 046d:c045 Logitech, USB-PS/2 Optical Mouse
	 low speed, power 98 mA, config 1, rev 27.20
	 driver: uhidev0
addr 04: 138a:0017 Validity Sensors, VFS5011 Fingerprint Reader
	 full speed, power 100 mA, config 1, rev 0.78, iSerial 5027128cc533
	 driver: ugen0
addr 05: 8087:07dc Intel, product 0x07dc
	 full speed, self powered, config 1, rev 0.01
	 driver: ugen1
addr 06: 5986:0268 SunplusIT INC., Integrated Camera
	 high speed, power 500 mA, config 1, rev 0.03
	 driver: uvideo0
Controller /dev/usb1:
addr 01: 8086:0000 Intel, EHCI root hub
	 high speed, self powered, config 1, rev 1.00
	 driver: uhub1
addr 02: 8087:8000 Intel, Rate Matching Hub
	 high speed, self powered, config 1, rev 0.04
	 driver: uhub3

```

and `xrandr` - I have a big screen (DP-1) hooked up on the mini-display port that runs over into the built-in laptop (monitor 0 / eDP-1) screen

```
└─ $ ▶ xrandr
Screen 0: minimum 320 x 200, current 1920 x 1848, maximum 16384 x 16384
eDP-1 connected primary 1366x768+0+1080 (normal left inverted right x axis y axis) 309mm x 173mm
   1366x768      60.06*+
   1280x720      60.00    59.99    59.86    59.74  
   1024x768      60.04    60.00  
   960x720       60.00  
   928x696       60.05  
   896x672       60.01  
   1024x576      59.95    59.96    59.90    59.82  
   960x600       59.93    60.00  
   960x540       59.96    59.99    59.63    59.82  
   800x600       60.00    60.32    56.25  
   840x525       60.01    59.88  
   864x486       59.92    59.57  
   700x525       59.98  
   800x450       59.95    59.82  
   640x512       60.02  
   700x450       59.96    59.88  
   640x480       60.00    59.94  
   720x405       59.51    58.99  
   684x384       59.88    59.85  
   640x400       59.88    59.98  
   640x360       59.86    59.83    59.84    59.32  
   512x384       60.00  
   512x288       60.00    59.92  
   480x270       59.63    59.82  
   400x300       60.32    56.34  
   432x243       59.92    59.57  
   320x240       60.05  
   360x202       59.51    59.13  
   320x180       59.84    59.32  
VGA-1 disconnected (normal left inverted right x axis y axis)
DP-1 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 527mm x 296mm
   1920x1080     60.00*+  74.97    50.00    59.94  
   1920x1080i    60.00    50.00    59.94  
   1600x1200     60.00  
   1680x1050     59.95  
   1280x1024     75.02    60.02  
   1440x900      59.89  
   1280x960      60.00  
   1152x864      75.00  
   1280x720      60.00    50.00    59.94  
   1024x768      75.03    60.00  
   832x624       74.55  
   800x600       75.00    60.32  
   720x576       50.00  
   720x480       60.00    59.94  
   640x480       75.00    60.00    59.94  
   720x400       70.08  
HDMI-1 disconnected (normal left inverted right x axis y axis)
DP-2 disconnected (normal left inverted right x axis y axis)
HDMI-2 disconnected (normal left inverted right x axis y axis)

```

dbwiddis commented:
USB output for your system shows it's parsing as expected.  Calling that one done.
```
 EHCI root hub (Intel)
 |-- Integrated Camera (SunplusIT INC.)
 |-- Rate Matching Hub (Intel)
 |-- USB-PS/2 Optical Mouse (Logitech)
 |-- VFS5011 Fingerprint Reader (Validity Sensors) [s/n: 5027128cc533]
 |-- product 0x07dc (Intel)
 EHCI root hub (Intel)
 |-- Rate Matching Hub (Intel)
```
For `xrandr` it needs the `--verbose` switch to get the EDID info. Or you can just run it on your system and tell me if a display is detected, which is sufficient.  If not, I have no other suggestions.

Enjoy your family time. No rush on this PR.  I may take a look at the filesystems later today... if it's just `mount`-based parsing I can figure it out.   Disks and partitions are the tough one... I think a combination of `hw.disknames` to get the list and `disklabel` is probably the best approach, but `disklabel` requires root permissions. 

One other note: on my VM, `mvn test` appears to go into an endless loop. Does it work for you?

mprins commented:
yeah, no problems using `mvn test` or `mvn install`.

mount provides quite a bit of logical info:
```
└─ $ ▶ mount -v
/dev/sd0a (2cf69345d371cd82.a) on / type ffs (rw, local, ctime=Wed Dec 30 09:25:50 2020)
/dev/sd0l (2cf69345d371cd82.l) on /home type ffs (rw, local, noatime, nodev, nosuid, softdep, ctime=Wed Dec 30 09:25:25 2020)
/dev/sd0d (2cf69345d371cd82.d) on /tmp type ffs (rw, local, nodev, nosuid, ctime=Wed Dec 30 09:25:25 2020)
/dev/sd0f (2cf69345d371cd82.f) on /usr type ffs (rw, local, nodev, ctime=Wed Dec 30 09:25:25 2020)
/dev/sd0g (2cf69345d371cd82.g) on /usr/X11R6 type ffs (rw, local, nodev, ctime=Wed Dec 30 09:25:25 2020)
/dev/sd0h (2cf69345d371cd82.h) on /usr/local type ffs (rw, local, nodev, wxallowed, ctime=Wed Dec 30 09:25:25 2020)
/dev/sd0k (2cf69345d371cd82.k) on /usr/obj type ffs (rw, local, nodev, nosuid, softdep, ctime=Wed Dec 30 09:25:25 2020)
/dev/sd0j (2cf69345d371cd82.j) on /usr/src type ffs (rw, local, nodev, nosuid, softdep, ctime=Wed Dec 30 09:25:25 2020)
/dev/sd0e (2cf69345d371cd82.e) on /var type ffs (rw, local, nodev, nosuid, ctime=Wed Dec 30 09:25:25 2020)
mfs:68662 on /var/www/dev type mfs (rw, asynchronous, local, nosuid, ctime=Wed Dec 30 09:25:25 2020, size=128 512-blocks)

```
but not the hardware info that disklabel does, 
```
└─ $ ▶ doas disklabel -v sd0  
doas (mprins@thinkpad.local) password: 
# /dev/rsd0c:
type: SCSI
disk: SCSI disk
label: SAMSUNG MZ7TD256
duid: 2cf69345d371cd82
flags:
bytes/sector: 512
sectors/track: 63
tracks/cylinder: 255
sectors/cylinder: 16065
cylinders: 31130
total sectors: 500118192
boundstart: 1024
boundend: 500103450
drivedata: 0 

16 partitions:
#                size           offset  fstype [fsize bsize   cpg]
  a:          2097152             1024  4.2BSD   2048 16384 12958 # /
  b:         17023368          2098176    swap                    # none
  c:        500118192                0  unused                    
  d:          8388576         19121568  4.2BSD   2048 16384 12958 # /tmp
  e:         41386752         27510144  4.2BSD   2048 16384 12958 # /var
  f:          4194304         68896896  4.2BSD   2048 16384 12958 # /usr
  g:          2097152         73091200  4.2BSD   2048 16384 12958 # /usr/X11R6
  h:         20971520         75188352  4.2BSD   2048 16384 12958 # /usr/local
  i:              960               64   MSDOS                    
  j:          4194304         96159872  4.2BSD   2048 16384 12958 # /usr/src
  k:         12582912        100354176  4.2BSD   2048 16384 12958 # /usr/obj
  l:        387166336        112937088  4.2BSD   4096 32768 26062 # /home

```
the `c` partition is the whole disk -by convention- , and the `i` partition is an UEFI partition in this case

```
└─ $ ▶ xrandr --verbose
Screen 0: minimum 320 x 200, current 1920 x 1848, maximum 16384 x 16384
eDP-1 connected primary 1366x768+0+1080 (0x49) normal (normal left inverted right x axis y axis) 309mm x 173mm
	Identifier: 0x42
	Timestamp:  558029
	Subpixel:   unknown
	Gamma:      1.0:1.0:1.0
	Brightness: 1.0
	Clones:    
	CRTC:       0
	CRTCs:      0 1 2
	Transform:  1.000000 0.000000 0.000000
	            0.000000 1.000000 0.000000
	            0.000000 0.000000 1.000000
	           filter: 
	EDID: 
		00ffffffffffff0006af3c3300000000
		00170104951f117802afe59658538a26
		24505400000001010101010101010101
		010101010101521c56a2500016302616
		360035ad100000180000000f00000000
		00000000000000000020000000fe0041
		554f0a202020202020202020000000fe
		004231343058544e30332e33200a00b9
	Backlight: 15 
		range: (0, 15)
	scaling mode: Full aspect 
		supported: Full, Center, Full aspect
	Colorspace: Default 
		supported: Default, RGB_Wide_Gamut_Fixed_Point, RGB_Wide_Gamut_Floating_Point, opRGB, DCI-P3_RGB_D65, BT2020_RGB, BT601_YCC, BT709_YCC, XVYCC_601, XVYCC_709, SYCC_601, opYCC_601, BT2020_CYCC, BT2020_YCC
	max bpc: 12 
		range: (6, 12)
	Broadcast RGB: Automatic 
		supported: Automatic, Full, Limited 16:235
	link-status: Good 
		supported: Good, Bad
	CONNECTOR_ID: 79 
		supported: 79
	non-desktop: 0 
		range: (0, 1)
  1366x768 (0x49) 72.500MHz -HSync -VSync *current +preferred
        h: width  1366 start 1404 end 1426 total 1528 skew    0 clock  47.45KHz
        v: height  768 start  771 end  777 total  790           clock  60.06Hz
  1280x720 (0x4a) 156.125MHz -HSync +VSync DoubleScan
        h: width  1280 start 1376 end 1512 total 1744 skew    0 clock  89.52KHz
        v: height  720 start  721 end  724 total  746           clock  60.00Hz
  1280x720 (0x4b) 120.750MHz +HSync -VSync DoubleScan
        h: width  1280 start 1304 end 1320 total 1360 skew    0 clock  88.79KHz
        v: height  720 start  721 end  724 total  740           clock  59.99Hz
  1280x720 (0x4c) 74.500MHz -HSync +VSync
        h: width  1280 start 1344 end 1472 total 1664 skew    0 clock  44.77KHz
        v: height  720 start  723 end  728 total  748           clock  59.86Hz
  1280x720 (0x4d) 63.750MHz +HSync -VSync
        h: width  1280 start 1328 end 1360 total 1440 skew    0 clock  44.27KHz
        v: height  720 start  723 end  728 total  741           clock  59.74Hz
  1024x768 (0x4e) 133.475MHz -HSync +VSync DoubleScan
        h: width  1024 start 1100 end 1212 total 1400 skew    0 clock  95.34KHz
        v: height  768 start  768 end  770 total  794           clock  60.04Hz
  1024x768 (0x4f) 65.000MHz -HSync -VSync
        h: width  1024 start 1048 end 1184 total 1344 skew    0 clock  48.36KHz
        v: height  768 start  771 end  777 total  806           clock  60.00Hz
  960x720 (0x50) 117.000MHz -HSync +VSync DoubleScan
        h: width   960 start 1024 end 1128 total 1300 skew    0 clock  90.00KHz
        v: height  720 start  720 end  722 total  750           clock  60.00Hz
  928x696 (0x51) 109.150MHz -HSync +VSync DoubleScan
        h: width   928 start  976 end 1088 total 1264 skew    0 clock  86.35KHz
        v: height  696 start  696 end  698 total  719           clock  60.05Hz
  896x672 (0x52) 102.400MHz -HSync +VSync DoubleScan
        h: width   896 start  960 end 1060 total 1224 skew    0 clock  83.66KHz
        v: height  672 start  672 end  674 total  697           clock  60.01Hz
  1024x576 (0x53) 98.500MHz -HSync +VSync DoubleScan
        h: width  1024 start 1092 end 1200 total 1376 skew    0 clock  71.58KHz
        v: height  576 start  577 end  580 total  597           clock  59.95Hz
  1024x576 (0x54) 78.375MHz +HSync -VSync DoubleScan
        h: width  1024 start 1048 end 1064 total 1104 skew    0 clock  70.99KHz
        v: height  576 start  577 end  580 total  592           clock  59.96Hz
  1024x576 (0x55) 46.500MHz -HSync +VSync
        h: width  1024 start 1064 end 1160 total 1296 skew    0 clock  35.88KHz
        v: height  576 start  579 end  584 total  599           clock  59.90Hz
  1024x576 (0x56) 42.000MHz +HSync -VSync
        h: width  1024 start 1072 end 1104 total 1184 skew    0 clock  35.47KHz
        v: height  576 start  579 end  584 total  593           clock  59.82Hz
  960x600 (0x57) 96.625MHz -HSync +VSync DoubleScan
        h: width   960 start 1028 end 1128 total 1296 skew    0 clock  74.56KHz
        v: height  600 start  601 end  604 total  622           clock  59.93Hz
  960x600 (0x58) 77.000MHz +HSync -VSync DoubleScan
        h: width   960 start  984 end 1000 total 1040 skew    0 clock  74.04KHz
        v: height  600 start  601 end  604 total  617           clock  60.00Hz
  960x540 (0x59) 86.500MHz -HSync +VSync DoubleScan
        h: width   960 start 1024 end 1124 total 1288 skew    0 clock  67.16KHz
        v: height  540 start  541 end  544 total  560           clock  59.96Hz
  960x540 (0x5a) 69.250MHz +HSync -VSync DoubleScan
        h: width   960 start  984 end 1000 total 1040 skew    0 clock  66.59KHz
        v: height  540 start  541 end  544 total  555           clock  59.99Hz
  960x540 (0x5b) 40.750MHz -HSync +VSync
        h: width   960 start  992 end 1088 total 1216 skew    0 clock  33.51KHz
        v: height  540 start  543 end  548 total  562           clock  59.63Hz
  960x540 (0x5c) 37.250MHz +HSync -VSync
        h: width   960 start 1008 end 1040 total 1120 skew    0 clock  33.26KHz
        v: height  540 start  543 end  548 total  556           clock  59.82Hz
  800x600 (0x5d) 81.000MHz +HSync +VSync DoubleScan
        h: width   800 start  832 end  928 total 1080 skew    0 clock  75.00KHz
        v: height  600 start  600 end  602 total  625           clock  60.00Hz
  800x600 (0x5e) 40.000MHz +HSync +VSync
        h: width   800 start  840 end  968 total 1056 skew    0 clock  37.88KHz
        v: height  600 start  601 end  605 total  628           clock  60.32Hz
  800x600 (0x5f) 36.000MHz +HSync +VSync
        h: width   800 start  824 end  896 total 1024 skew    0 clock  35.16KHz
        v: height  600 start  601 end  603 total  625           clock  56.25Hz
  840x525 (0x60) 73.125MHz -HSync +VSync DoubleScan
        h: width   840 start  892 end  980 total 1120 skew    0 clock  65.29KHz
        v: height  525 start  526 end  529 total  544           clock  60.01Hz
  840x525 (0x61) 59.500MHz +HSync -VSync DoubleScan
        h: width   840 start  864 end  880 total  920 skew    0 clock  64.67KHz
        v: height  525 start  526 end  529 total  540           clock  59.88Hz
  864x486 (0x62) 32.500MHz -HSync +VSync
        h: width   864 start  888 end  968 total 1072 skew    0 clock  30.32KHz
        v: height  486 start  489 end  494 total  506           clock  59.92Hz
  864x486 (0x63) 30.500MHz +HSync -VSync
        h: width   864 start  912 end  944 total 1024 skew    0 clock  29.79KHz
        v: height  486 start  489 end  494 total  500           clock  59.57Hz
  700x525 (0x64) 61.000MHz +HSync +VSync DoubleScan
        h: width   700 start  744 end  820 total  940 skew    0 clock  64.89KHz
        v: height  525 start  526 end  532 total  541           clock  59.98Hz
  800x450 (0x65) 59.125MHz -HSync +VSync DoubleScan
        h: width   800 start  848 end  928 total 1056 skew    0 clock  55.99KHz
        v: height  450 start  451 end  454 total  467           clock  59.95Hz
  800x450 (0x66) 48.750MHz +HSync -VSync DoubleScan
        h: width   800 start  824 end  840 total  880 skew    0 clock  55.40KHz
        v: height  450 start  451 end  454 total  463           clock  59.82Hz
  640x512 (0x67) 54.000MHz +HSync +VSync DoubleScan
        h: width   640 start  664 end  720 total  844 skew    0 clock  63.98KHz
        v: height  512 start  512 end  514 total  533           clock  60.02Hz
  700x450 (0x68) 51.750MHz -HSync +VSync DoubleScan
        h: width   700 start  740 end  812 total  924 skew    0 clock  56.01KHz
        v: height  450 start  451 end  456 total  467           clock  59.96Hz
  700x450 (0x69) 43.250MHz +HSync -VSync DoubleScan
        h: width   700 start  724 end  740 total  780 skew    0 clock  55.45KHz
        v: height  450 start  451 end  456 total  463           clock  59.88Hz
  640x480 (0x6a) 54.000MHz +HSync +VSync DoubleScan
        h: width   640 start  688 end  744 total  900 skew    0 clock  60.00KHz
        v: height  480 start  480 end  482 total  500           clock  60.00Hz
  640x480 (0x6b) 25.175MHz -HSync -VSync
        h: width   640 start  656 end  752 total  800 skew    0 clock  31.47KHz
        v: height  480 start  490 end  492 total  525           clock  59.94Hz
  720x405 (0x6c) 22.500MHz -HSync +VSync
        h: width   720 start  744 end  808 total  896 skew    0 clock  25.11KHz
        v: height  405 start  408 end  413 total  422           clock  59.51Hz
  720x405 (0x6d) 21.750MHz +HSync -VSync
        h: width   720 start  768 end  800 total  880 skew    0 clock  24.72KHz
        v: height  405 start  408 end  413 total  419           clock  58.99Hz
  684x384 (0x6e) 42.625MHz -HSync +VSync DoubleScan
        h: width   684 start  720 end  788 total  892 skew    0 clock  47.79KHz
        v: height  384 start  385 end  390 total  399           clock  59.88Hz
  684x384 (0x6f) 36.125MHz +HSync -VSync DoubleScan
        h: width   684 start  708 end  724 total  764 skew    0 clock  47.28KHz
        v: height  384 start  385 end  390 total  395           clock  59.85Hz
  640x400 (0x70) 41.750MHz -HSync +VSync DoubleScan
        h: width   640 start  676 end  740 total  840 skew    0 clock  49.70KHz
        v: height  400 start  401 end  404 total  415           clock  59.88Hz
  640x400 (0x71) 35.500MHz +HSync -VSync DoubleScan
        h: width   640 start  664 end  680 total  720 skew    0 clock  49.31KHz
        v: height  400 start  401 end  404 total  411           clock  59.98Hz
  640x360 (0x72) 37.250MHz -HSync +VSync DoubleScan
        h: width   640 start  672 end  736 total  832 skew    0 clock  44.77KHz
        v: height  360 start  361 end  364 total  374           clock  59.86Hz
  640x360 (0x73) 31.875MHz +HSync -VSync DoubleScan
        h: width   640 start  664 end  680 total  720 skew    0 clock  44.27KHz
        v: height  360 start  361 end  364 total  370           clock  59.83Hz
  640x360 (0x74) 18.000MHz -HSync +VSync
        h: width   640 start  664 end  720 total  800 skew    0 clock  22.50KHz
        v: height  360 start  363 end  368 total  376           clock  59.84Hz
  640x360 (0x75) 17.750MHz +HSync -VSync
        h: width   640 start  688 end  720 total  800 skew    0 clock  22.19KHz
        v: height  360 start  363 end  368 total  374           clock  59.32Hz
  512x384 (0x76) 32.500MHz -HSync -VSync DoubleScan
        h: width   512 start  524 end  592 total  672 skew    0 clock  48.36KHz
        v: height  384 start  385 end  388 total  403           clock  60.00Hz
  512x288 (0x77) 23.250MHz -HSync +VSync DoubleScan
        h: width   512 start  532 end  580 total  648 skew    0 clock  35.88KHz
        v: height  288 start  289 end  292 total  299           clock  60.00Hz
  512x288 (0x78) 21.000MHz +HSync -VSync DoubleScan
        h: width   512 start  536 end  552 total  592 skew    0 clock  35.47KHz
        v: height  288 start  289 end  292 total  296           clock  59.92Hz
  480x270 (0x79) 20.375MHz -HSync +VSync DoubleScan
        h: width   480 start  496 end  544 total  608 skew    0 clock  33.51KHz
        v: height  270 start  271 end  274 total  281           clock  59.63Hz
  480x270 (0x7a) 18.625MHz +HSync -VSync DoubleScan
        h: width   480 start  504 end  520 total  560 skew    0 clock  33.26KHz
        v: height  270 start  271 end  274 total  278           clock  59.82Hz
  400x300 (0x7b) 20.000MHz +HSync +VSync DoubleScan
        h: width   400 start  420 end  484 total  528 skew    0 clock  37.88KHz
        v: height  300 start  300 end  302 total  314           clock  60.32Hz
  400x300 (0x7c) 18.000MHz +HSync +VSync DoubleScan
        h: width   400 start  412 end  448 total  512 skew    0 clock  35.16KHz
        v: height  300 start  300 end  301 total  312           clock  56.34Hz
  432x243 (0x7d) 16.250MHz -HSync +VSync DoubleScan
        h: width   432 start  444 end  484 total  536 skew    0 clock  30.32KHz
        v: height  243 start  244 end  247 total  253           clock  59.92Hz
  432x243 (0x7e) 15.250MHz +HSync -VSync DoubleScan
        h: width   432 start  456 end  472 total  512 skew    0 clock  29.79KHz
        v: height  243 start  244 end  247 total  250           clock  59.57Hz
  320x240 (0x7f) 12.587MHz -HSync -VSync DoubleScan
        h: width   320 start  328 end  376 total  400 skew    0 clock  31.47KHz
        v: height  240 start  245 end  246 total  262           clock  60.05Hz
  360x202 (0x80) 11.250MHz -HSync +VSync DoubleScan
        h: width   360 start  372 end  404 total  448 skew    0 clock  25.11KHz
        v: height  202 start  204 end  206 total  211           clock  59.51Hz
  360x202 (0x81) 10.875MHz +HSync -VSync DoubleScan
        h: width   360 start  384 end  400 total  440 skew    0 clock  24.72KHz
        v: height  202 start  204 end  206 total  209           clock  59.13Hz
  320x180 (0x82)  9.000MHz -HSync +VSync DoubleScan
        h: width   320 start  332 end  360 total  400 skew    0 clock  22.50KHz
        v: height  180 start  181 end  184 total  188           clock  59.84Hz
  320x180 (0x83)  8.875MHz +HSync -VSync DoubleScan
        h: width   320 start  344 end  360 total  400 skew    0 clock  22.19KHz
        v: height  180 start  181 end  184 total  187           clock  59.32Hz
VGA-1 disconnected (normal left inverted right x axis y axis)
	Identifier: 0x43
	Timestamp:  558029
	Subpixel:   unknown
	Clones:    
	CRTCs:      0 1 2
	Transform:  1.000000 0.000000 0.000000
	            0.000000 1.000000 0.000000
	            0.000000 0.000000 1.000000
	           filter: 
	link-status: Good 
		supported: Good, Bad
	CONNECTOR_ID: 76 
		supported: 76
	non-desktop: 0 
		range: (0, 1)
DP-1 connected 1920x1080+0+0 (0x84) normal (normal left inverted right x axis y axis) 527mm x 296mm
	Identifier: 0x44
	Timestamp:  558029
	Subpixel:   unknown
	Gamma:      1.0:1.0:1.0
	Brightness: 1.0
	Clones:     HDMI-1
	CRTC:       1
	CRTCs:      0 1 2
	Transform:  1.000000 0.000000 0.000000
	            0.000000 1.000000 0.000000
	            0.000000 0.000000 1.000000
	           filter: 
	EDID: 
		00ffffffffffff0026cd286110060000
		291b0104a5351e783aa1a5a755529f27
		125054a56b80710081408180a940b300
		9500d1c00101023a801871382d40582c
		45000f282100001e000000ff00313135
		36353734313231353532000000fd0037
		4c1e5312000a202020202020000000fc
		0058323438332f323438310a202001ea
		02031cf1830100004f90020304050607
		011112131415161f23090707023a8018
		71382d40582c45000f282100001f011d
		8018711c1620582c25000f282100009f
		011d007251d01e206e2855000f282100
		001e2a4480a070382740302035000f28
		2100001a000000000000000000000000
		0000000000000000000000000000007a
	Colorspace: Default 
		supported: Default, RGB_Wide_Gamut_Fixed_Point, RGB_Wide_Gamut_Floating_Point, opRGB, DCI-P3_RGB_D65, BT2020_RGB, BT601_YCC, BT709_YCC, XVYCC_601, XVYCC_709, SYCC_601, opYCC_601, BT2020_CYCC, BT2020_YCC
	max bpc: 12 
		range: (6, 12)
	Broadcast RGB: Automatic 
		supported: Automatic, Full, Limited 16:235
	audio: auto 
		supported: force-dvi, off, auto, on
	link-status: Good 
		supported: Good, Bad
	CONNECTOR_ID: 86 
		supported: 86
	non-desktop: 0 
		range: (0, 1)
  1920x1080 (0x84) 148.500MHz +HSync +VSync *current +preferred
        h: width  1920 start 2008 end 2052 total 2200 skew    0 clock  67.50KHz
        v: height 1080 start 1084 end 1089 total 1125           clock  60.00Hz
  1920x1080 (0x85) 174.500MHz +HSync -VSync
        h: width  1920 start 1968 end 2000 total 2080 skew    0 clock  83.89KHz
        v: height 1080 start 1083 end 1088 total 1119           clock  74.97Hz
  1920x1080 (0x86) 148.500MHz +HSync +VSync
        h: width  1920 start 2448 end 2492 total 2640 skew    0 clock  56.25KHz
        v: height 1080 start 1084 end 1089 total 1125           clock  50.00Hz
  1920x1080 (0x87) 148.352MHz +HSync +VSync
        h: width  1920 start 2008 end 2052 total 2200 skew    0 clock  67.43KHz
        v: height 1080 start 1084 end 1089 total 1125           clock  59.94Hz
  1920x1080i (0x88) 74.250MHz +HSync +VSync Interlace
        h: width  1920 start 2008 end 2052 total 2200 skew    0 clock  33.75KHz
        v: height 1080 start 1084 end 1094 total 1125           clock  60.00Hz
  1920x1080i (0x89) 74.250MHz +HSync +VSync Interlace
        h: width  1920 start 2448 end 2492 total 2640 skew    0 clock  28.12KHz
        v: height 1080 start 1084 end 1094 total 1125           clock  50.00Hz
  1920x1080i (0x8a) 74.176MHz +HSync +VSync Interlace
        h: width  1920 start 2008 end 2052 total 2200 skew    0 clock  33.72KHz
        v: height 1080 start 1084 end 1094 total 1125           clock  59.94Hz
  1600x1200 (0x8b) 162.000MHz +HSync +VSync
        h: width  1600 start 1664 end 1856 total 2160 skew    0 clock  75.00KHz
        v: height 1200 start 1201 end 1204 total 1250           clock  60.00Hz
  1680x1050 (0x8c) 146.250MHz -HSync +VSync
        h: width  1680 start 1784 end 1960 total 2240 skew    0 clock  65.29KHz
        v: height 1050 start 1053 end 1059 total 1089           clock  59.95Hz
  1280x1024 (0x8d) 135.000MHz +HSync +VSync
        h: width  1280 start 1296 end 1440 total 1688 skew    0 clock  79.98KHz
        v: height 1024 start 1025 end 1028 total 1066           clock  75.02Hz
  1280x1024 (0x8e) 108.000MHz +HSync +VSync
        h: width  1280 start 1328 end 1440 total 1688 skew    0 clock  63.98KHz
        v: height 1024 start 1025 end 1028 total 1066           clock  60.02Hz
  1440x900 (0x8f) 106.500MHz -HSync +VSync
        h: width  1440 start 1520 end 1672 total 1904 skew    0 clock  55.93KHz
        v: height  900 start  903 end  909 total  934           clock  59.89Hz
  1280x960 (0x90) 108.000MHz +HSync +VSync
        h: width  1280 start 1376 end 1488 total 1800 skew    0 clock  60.00KHz
        v: height  960 start  961 end  964 total 1000           clock  60.00Hz
  1152x864 (0x91) 108.000MHz +HSync +VSync
        h: width  1152 start 1216 end 1344 total 1600 skew    0 clock  67.50KHz
        v: height  864 start  865 end  868 total  900           clock  75.00Hz
  1280x720 (0x92) 74.250MHz +HSync +VSync
        h: width  1280 start 1390 end 1430 total 1650 skew    0 clock  45.00KHz
        v: height  720 start  725 end  730 total  750           clock  60.00Hz
  1280x720 (0x93) 74.250MHz +HSync +VSync
        h: width  1280 start 1720 end 1760 total 1980 skew    0 clock  37.50KHz
        v: height  720 start  725 end  730 total  750           clock  50.00Hz
  1280x720 (0x94) 74.176MHz +HSync +VSync
        h: width  1280 start 1390 end 1430 total 1650 skew    0 clock  44.96KHz
        v: height  720 start  725 end  730 total  750           clock  59.94Hz
  1024x768 (0x95) 78.750MHz +HSync +VSync
        h: width  1024 start 1040 end 1136 total 1312 skew    0 clock  60.02KHz
        v: height  768 start  769 end  772 total  800           clock  75.03Hz
  1024x768 (0x4f) 65.000MHz -HSync -VSync
        h: width  1024 start 1048 end 1184 total 1344 skew    0 clock  48.36KHz
        v: height  768 start  771 end  777 total  806           clock  60.00Hz
  832x624 (0x96) 57.284MHz -HSync -VSync
        h: width   832 start  864 end  928 total 1152 skew    0 clock  49.73KHz
        v: height  624 start  625 end  628 total  667           clock  74.55Hz
  800x600 (0x97) 49.500MHz +HSync +VSync
        h: width   800 start  816 end  896 total 1056 skew    0 clock  46.88KHz
        v: height  600 start  601 end  604 total  625           clock  75.00Hz
  800x600 (0x5e) 40.000MHz +HSync +VSync
        h: width   800 start  840 end  968 total 1056 skew    0 clock  37.88KHz
        v: height  600 start  601 end  605 total  628           clock  60.32Hz
  720x576 (0x98) 27.000MHz -HSync -VSync
        h: width   720 start  732 end  796 total  864 skew    0 clock  31.25KHz
        v: height  576 start  581 end  586 total  625           clock  50.00Hz
  720x480 (0x99) 27.027MHz -HSync -VSync
        h: width   720 start  736 end  798 total  858 skew    0 clock  31.50KHz
        v: height  480 start  489 end  495 total  525           clock  60.00Hz
  720x480 (0x9a) 27.000MHz -HSync -VSync
        h: width   720 start  736 end  798 total  858 skew    0 clock  31.47KHz
        v: height  480 start  489 end  495 total  525           clock  59.94Hz
  640x480 (0x9b) 31.500MHz -HSync -VSync
        h: width   640 start  656 end  720 total  840 skew    0 clock  37.50KHz
        v: height  480 start  481 end  484 total  500           clock  75.00Hz
  640x480 (0x9c) 25.200MHz -HSync -VSync
        h: width   640 start  656 end  752 total  800 skew    0 clock  31.50KHz
        v: height  480 start  490 end  492 total  525           clock  60.00Hz
  640x480 (0x6b) 25.175MHz -HSync -VSync
        h: width   640 start  656 end  752 total  800 skew    0 clock  31.47KHz
        v: height  480 start  490 end  492 total  525           clock  59.94Hz
  720x400 (0x9d) 28.320MHz -HSync +VSync
        h: width   720 start  738 end  846 total  900 skew    0 clock  31.47KHz
        v: height  400 start  412 end  414 total  449           clock  70.08Hz
HDMI-1 disconnected (normal left inverted right x axis y axis)
	Identifier: 0x45
	Timestamp:  558029
	Subpixel:   unknown
	Clones:     DP-1
	CRTCs:      0 1 2
	Transform:  1.000000 0.000000 0.000000
	            0.000000 1.000000 0.000000
	            0.000000 0.000000 1.000000
	           filter: 
	max bpc: 12 
		range: (8, 12)
	content type: No Data 
		supported: No Data, Graphics, Photo, Cinema, Game
	Colorspace: Default 
		supported: Default, SMPTE_170M_YCC, BT709_YCC, XVYCC_601, XVYCC_709, SYCC_601, opYCC_601, opRGB, BT2020_CYCC, BT2020_RGB, BT2020_YCC, DCI-P3_RGB_D65, DCI-P3_RGB_Theater
	aspect ratio: Automatic 
		supported: Automatic, 4:3, 16:9
	Broadcast RGB: Automatic 
		supported: Automatic, Full, Limited 16:235
	audio: auto 
		supported: force-dvi, off, auto, on
	link-status: Good 
		supported: Good, Bad
	CONNECTOR_ID: 93 
		supported: 93
	non-desktop: 0 
		range: (0, 1)
DP-2 disconnected (normal left inverted right x axis y axis)
	Identifier: 0x46
	Timestamp:  558029
	Subpixel:   unknown
	Clones:     HDMI-2
	CRTCs:      0 1 2
	Transform:  1.000000 0.000000 0.000000
	            0.000000 1.000000 0.000000
	            0.000000 0.000000 1.000000
	           filter: 
	Colorspace: Default 
		supported: Default, RGB_Wide_Gamut_Fixed_Point, RGB_Wide_Gamut_Floating_Point, opRGB, DCI-P3_RGB_D65, BT2020_RGB, BT601_YCC, BT709_YCC, XVYCC_601, XVYCC_709, SYCC_601, opYCC_601, BT2020_CYCC, BT2020_YCC
	max bpc: 12 
		range: (6, 12)
	Broadcast RGB: Automatic 
		supported: Automatic, Full, Limited 16:235
	audio: auto 
		supported: force-dvi, off, auto, on
	link-status: Good 
		supported: Good, Bad
	CONNECTOR_ID: 99 
		supported: 99
	non-desktop: 0 
		range: (0, 1)
HDMI-2 disconnected (normal left inverted right x axis y axis)
	Identifier: 0x47
	Timestamp:  558029
	Subpixel:   unknown
	Clones:     DP-2
	CRTCs:      0 1 2
	Transform:  1.000000 0.000000 0.000000
	            0.000000 1.000000 0.000000
	            0.000000 0.000000 1.000000
	           filter: 
	max bpc: 12 
		range: (8, 12)
	content type: No Data 
		supported: No Data, Graphics, Photo, Cinema, Game
	Colorspace: Default 
		supported: Default, SMPTE_170M_YCC, BT709_YCC, XVYCC_601, XVYCC_709, SYCC_601, opYCC_601, opRGB, BT2020_CYCC, BT2020_RGB, BT2020_YCC, DCI-P3_RGB_D65, DCI-P3_RGB_Theater
	aspect ratio: Automatic 
		supported: Automatic, 4:3, 16:9
	Broadcast RGB: Automatic 
		supported: Automatic, Full, Limited 16:235
	audio: auto 
		supported: force-dvi, off, auto, on
	link-status: Good 
		supported: Good, Bad
	CONNECTOR_ID: 105 
		supported: 105
	non-desktop: 0 
		range: (0, 1)

```

dbwiddis commented:
Looks like `mount` just has an extra field: `(2cf69345d371cd82.a)` etc. so should be an easy fix.  I'll throw something together today (and probably combine duplicated `mount` code elsewhere in the project in the process.

The verbose `xrandr` looks like it supplies the needed info:
```
	EDID: 
		00ffffffffffff0006af3c3300000000
		00170104951f117802afe59658538a26
... etc
```
So we should be able to copy the existing code from all the other Unix classes. Linux even duplicates the `XRandr` class (so should be trimmed to conform). 

`disklabel` is definitely the way to go with root permissions. Also `systat iostat` may have some info (again, root). The challenge is piecing together what we can without root access.  That's still your assignment for this week. :)

dbwiddis commented:
`OpenBsdFileSystem` (and `OpenBsdOSFileStore`) is done(-ish).  It was easy to copy from the Aix version and just change the parsing of `df -i` and `mount -v`.   I'll need you to go through and check a few things:
 - I wasn't sure what to do with the second column: `(2cf69345d371cd82.a)`.  Currently I have `sd0a` for both name and label, and that field as the uuid field, but as a user you may prefer to change how they're assigned.
 - Is `tmpfs` a thing in OpenBSD, creating a Ram Disk?  Check out the "description" logic and tweak as necessary.

Also, earlier, I combined all the X Display code into a single common Unix class so that's done.

Down to disks and partitions!  (And all the fun of trying to sort out "logical partitions" which the OSHI API isn't very clear about.)

dbwiddis commented:
Taking a look at the ComputerSystem + Firmware + Baseboard:
 - Since this information isn't going to change, use the `Memoizer` as a lazy getter calling static query methods.  See `LinuxComputerSystem` for a template to follow (the query methods will be simpler).  In a future, more modular major version, those static query methods will be moved into classes in the `driver` packages as we continue to try to separate API and implementation. You could put them there now (perhaps a `SysctlHw` class) if you want or just leave in the same class.
 - You've duplicated `hw.serialno` for both ComputerSystem and Baseboard.  Is there opportunity to store `hw.uuid` in the Baseboard serial number instead?
 - I see the third line of your `bios0` output in `dmesg` is: `bios0: LENOVO 20AWA08J00`.  This info is in `hw.vendor` and `hw.product` which I see you've used for baseboard.  However, I think `20AWA08J00` is a model, so perhaps that might be good to combine with your `hw.version` in the ComputerSystem model field, e.g., "ThinkPad T440p (20AWA08J00)".  I'll leave that choice to you, though. It's probably fine to have that `hw.product` value in the baseboard field as you have done... but see next bullet.
 - The Baseboard "version" field duplicates the ComputerSystem "model" field. This doesn't seem right (at best you should put the `hw.product` value as version and `hw.version` value as model to be consistent.)

dbwiddis commented:
Actually ignore that uuid request. We should really have a new field on the ComputerSystem object with that!  I'll just add a new issue to do that later.

dbwiddis commented:
I added the HW UUID in another PR and just updated this one as the last port.  Also I tweaked the other comments I made above (using memoizers and swapping the product and version for the baseboard).

dbwiddis commented on oshi-core/src/main/java/oshi/driver/unix/openbsd/disk/Disklabel.java:
```diff
@@ -0,0 +1,127 @@
+                }
+                split = line.split(":");
+                if (split.length == 2) {
+                    params.put(split[0], split[1].trim());
```
I don't see a need to create a map just to store the one value you need.  Just create a `bytesPerSector` variable, and populate it when you see that text with the value from `ParseUtil.getFirstInt()`.

dbwiddis commented on oshi-core/src/main/java/oshi/driver/unix/openbsd/disk/Disklabel.java:
```diff
@@ -0,0 +1,127 @@
+                //  l:        387166336        112937088  4.2BSD   4096 32768 26062 # /home
+                // Note size is in sectors
+                if (line.startsWith("#") || line.isEmpty()) {
+                    continue;
```
Style preference. Some of our CI complains if there are too many "break" and "continue" statements.  So we could wrap the code below this inside an 'if'.   I'm thinking we could also ignore the previous loop advancing to the "16 partitions" and simply perform a check for `line.indexOf(':')` which should always be 2 (untrimmed) or 1 (trimmed) and return only these lines.  That also allows using `substring(2)` instead of the split on `:`.

dbwiddis commented on oshi-core/src/main/java/oshi/driver/unix/openbsd/disk/Disklabel.java:
```diff
@@ -0,0 +1,127 @@
+        // partition, and the `b' partition of the boot disk is the swap partition
+        // List<String> cmd = ExecutingCommand.runNative("doas disklabel -n " + diskName);
+        List<String> cmd = ExecutingCommand.runNative("disklabel -n " + diskName);
+        if (cmd.size() > 0) {
```
`!cmd.isEmpty()` is preferred, although you can usually just iterate over the list and if it's empty nothing will happen inside the list.

dbwiddis commented on oshi-core/src/main/java/oshi/driver/unix/openbsd/disk/Disklabel.java:
```diff
@@ -0,0 +1,127 @@
+                    continue;
+                }
+                split = ParseUtil.whitespaces.split(line.trim(), 9);
+                name = split[0].replaceAll(":", "");
```
Style/performance preference: Both splitting and `replaceAll` use regex.  See above on what I think is a better option using `substring()`.

dbwiddis commented on oshi-core/src/main/java/oshi/driver/unix/openbsd/disk/Disklabel.java:
```diff
@@ -0,0 +1,127 @@
+                split = ParseUtil.whitespaces.split(line.trim(), 9);
+                name = split[0].replaceAll(":", "");
+                if (split.length > 5) {
+                    p = new HWPartition(diskName + name, name, split[3], Constants.UNKNOWN,
```
Style preference: don't like this whole line being repeated twice, when the only difference appears to be the last parameter.  Seems that could be handled with a ternary argument inline to reduce this to a single line, e.g., 
```
if (split.length > 3) {
   partitions.add(new HWPartition( ... stuff ..., split.length > 4 ? split[split.length-1] : ""));
}
```

dbwiddis commented on oshi-core/src/main/java/oshi/hardware/platform/unix/openbsd/OpenBsdHWDiskStore.java:
```diff
@@ -62,6 +57,68 @@ private OpenBsdHWDiskStore(String name, String model, String serial, long size)
+                    diskdata.getA().get("label"),
+                    diskdata.getA().get("label"),
+                    diskdata.getA().get("duid"),
+                    ParseUtil.parseLongOrDefault(diskdata.getA().get("bytes/sector"), 0) * ParseUtil
```
Something just doesn't seem right using a Pair and then a map with String Values that have to be parsed and then later multiplied together.   Could this not all be a single Quintet or even just a class with sensible names and proper variable typing internally?  Alternately, if the "C" partition is the "whole disk" could you not just send back that partition separately with all these values populated so you'd have a `Pair<HWPartition, List<HWPartition>` where `getA()` is the whole-disk stuff and `getB()` is everything else?

dbwiddis commented on oshi-core/src/main/java/oshi/hardware/platform/unix/openbsd/OpenBsdHWDiskStore.java:
```diff
@@ -62,6 +57,68 @@ private OpenBsdHWDiskStore(String name, String model, String serial, long size)
+                .getDiskParams((device.split(":")[0]));
+            if (diskdata.getA().isEmpty()) {
+                // something must have gone wrong, probably no elevated/root permission
+                store = new OpenBsdHWDiskStore((device.split(":")[0]), Constants.UNKNOWN, Constants.UNKNOWN, 0L);
```
Do we have a backup non-root source of at least some partition info, particularly for mounted partitions?  I think `df` or `iostat` probably can at least help here.

dbwiddis reviewed (commented):
Some comments and suggestions!

dbwiddis commented on oshi-core/src/main/java/oshi/driver/unix/openbsd/disk/Disklabel.java:
```diff
@@ -0,0 +1,127 @@
+                } else if (split.length == 4) {
+                    p = new HWPartition(diskName + name, name, split[3], Constants.UNKNOWN,
+                        ParseUtil.parseLongOrDefault(split[1], -1) * ParseUtil
+                            .parseLongOrDefault(params.get("bytes/sector"), -1), -1, -1, "");
```
One more note: you're using -1 for major and minor. I believe these values are accessible as the filesize of the device.  See `ls -al` output for /dev for the given partition to see what I mean.  I think you can extract it using `stat` -- see the code in `oshi.driver.unix.freebsd.disk.GeomPartList.java`.  If root permissions don't allow that, you can see it in `ls` output.

dbwiddis commented:
Made some tweaks!  Let me know what you think!  please try with/without root.

Are we done here?

dbwiddis commented:
OK, not quite done.  For non-root HW disk store I think we can parse `dmesg` for disk size:
```
wd0 at pciide0 channel 0 drive 0: <OpenBSD-0 SSD>
wd0: 128-sector PIO, LBA48, 8192MB, 16777216 sectors
```

dbwiddis commented:
I think this is ready to merge. Anything else you think needs doing?

dbwiddis commented:
Heh, I just fixed those spaces and tried to push and it yelled at me.  Thanks. :)

mprins commented:
attached results of running `mvn clean test -pl :oshi-core -Dtest=SystemInfoTest` (regular user)
[oshi.SystemInfoTest-output.txt](https://github.com/oshi/oshi/files/5818903/oshi.SystemInfoTest-output.txt)
and  `doas mvn clean test -pl :oshi-core -Dtest=SystemInfoTest` (elevated priviledges) 
[doas-oshi.SystemInfoTest-output.txt](https://github.com/oshi/oshi/files/5818931/doas-oshi.SystemInfoTest-output.txt)

So I think all the functionality is there, I still see a test failure while running install:
```
[ERROR] Tests run: 10, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 101.161 s <<< FAILURE! - in oshi.software.os.OperatingSystemTest
[ERROR] oshi.software.os.OperatingSystemTest.testProcessStats  Time elapsed: 0.185 s  <<< FAILURE!
java.lang.AssertionError: 
Current running process thread count should be greater than 0
Expected: is a value greater than <0>
     but: <0> was equal to <0>
	at oshi.software.os.OperatingSystemTest.testProcessStats(OperatingSystemTest.java:119)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   OperatingSystemTest.testProcessStats:119 Current running process thread count should be greater than 0
Expected: is a value greater than <0>
     but: <0> was equal to <0>

```

And on a vbox vm the cpu stuff was odd yesterday, but may also be caused by vbox. 
Off to work now

dbwiddis commented:
Hmm, interesting with non-root you still don't get the disk model and size.  Can you post your `dmesg` output for lines beginning with `sd0`?

As for the test error, apparently thread count is 0.  This may be a race condition for a process which ended before `updateThreadCount()` was called.  Should probably have the default set to 1 on line 349, rather than 0.

mprins commented:
I'll look up my laptop dmesg tonight.

here's the vbox oddity...

```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   CentralProcessorTest.testFrequencies:83 Central Processor's logical processor frequency array length should be the same as its logical processor count
Expected: is <3>
     but: was <1>
[INFO] 
[ERROR] Tests run: 116, Failures: 1, Errors: 0, Skipped: 0

```
while `sysctl hw` says we have both 1 and 3 processors ;-)
```
hw.machine=amd64
hw.model=AMD Ryzen 7 3700X 8-Core Processor
hw.ncpu=1
hw.byteorder=1234
hw.pagesize=4096
hw.disknames=wd0:485ae173f5008c6c,cd0:
hw.diskcount=2
hw.sensors.acpiac0.indicator0=On (power supply)
hw.cpuspeed=3600
hw.vendor=innotek GmbH
hw.product=VirtualBox
hw.version=1.2
hw.serialno=0
hw.uuid=526d3985-a79f-fc41-b3d2-3687111ac93c
hw.physmem=4278124544
hw.usermem=4278108160
hw.ncpufound=3
hw.allowpowerdown=1
hw.smt=0
hw.ncpuonline=1
```
while `dmesg` reports just 1 cpu:
```
cpu0 at mainbus0: apid 0 (boot processor)
cpu0: AMD Ryzen 7 3700X 8-Core Processor, 3600.55 MHz, 17-71-00
cpu0: FPU,VME,DE,PSE,TSC,MSR,PAE,MCE,CX8,APIC,SEP,MTRR,PGE,MCA,CMOV,PAT,PSE36,CFLUSH,MMX,FXSR,SSE,SSE2,HTT,SSE3,PCLMUL,SSSE3,CX16,SSE4.1,SSE4.2,MOVBE,POPCNT,AES,XSAVE,AVX,RDRAND,NXE,MMXX,FFXSR,RDTSCP,LONG,LAHF,CMPLEG,SVM,AMCR8,ABM,SSE4A,MASSE,3DNOWP,ITSC,FSGSBASE,AVX2,RDSEED,CLFLUSHOPT
cpu0: 32KB 64b/line 8-way I-cache, 32KB 64b/line 8-way D-cache, 512KB 64b/line 8-way L2 cache, 32MB 64b/line disabled L3 cache
cpu0: ITLB 64 4KB entries fully associative, 64 4MB entries fully associative
cpu0: DTLB 64 4KB entries fully associative, 64 4MB entries fully associative
cpu0: smt 0, core 0, package 0
mtrr: CPU supports MTRRs but not enabled by BIOS
cpu0: apic clock running at 1000MHz
cpu at mainbus0: not configured
cpu at mainbus0: not configured

```
perhaps we should go with the number reported by `hw.ncpuonline`?

dbwiddis commented:
> perhaps we should go with the number reported by hw.ncpuonline

Does that change during operation?  Why do we keep fetching the sysctl value when getting cpu frequency?  (And per earlier comment rather than using modulo 2 we should line up with smt threads from dmesg if we continue to do it that way, which I'm not sure we should.)

Whichever we choose we should pick the same one for counting CPUs as for cpu frequency... we can just fill an array with `getLogicalProcessorCount()` entries.   The question becomes which is a more accurate statement of the number of processors.

In this case, it looks like the vm is running the single-threaded kernel and not `bsd.mp` which might be able to see the other processors.   We really don't have any sort of variable anywhere that says "this hardware exists but the OS can't see it".

dbwiddis commented:
Also I found another format for disk size in another dmesg so I'll try to add that.
```
sd0 at scsibus1 targ 0 lun 0: <ATA, P3-2TB, T180> SCSI3 0/direct fixed naa.5000000000000000
sd0: 1953514MB, 512 bytes/sector, 4000797360 sectors, thin
sd1 at scsibus1 targ 2 lun 0: <ATA, MT-1TB, T180> SCSI3 0/direct fixed naa.5000000000000000
sd1: 976762MB, 512 bytes/sector, 2000409264 sectors, thin
```

dbwiddis commented:
So I think we're all done now except for determining the array size of the cpu frequency array, and whether to fill it all with the same number or to be complex and figure out which might be smt threads and set them to 0.

My preference is to just go simple.  Create an array of size `getLogicalProcessorCount()` and fill it with the same value.  [That's what we do on macOS](https://github.com/oshi/oshi/blob/master/oshi-core/src/main/java/oshi/hardware/platform/mac/MacCentralProcessor.java#L156).

mprins commented:
> > perhaps we should go with the number reported by hw.ncpuonline
> 
> Does that change during operation? Why do we keep fetching the sysctl value when getting cpu frequency? (And per earlier comment rather than using modulo 2 we should line up with smt threads from dmesg if we continue to do it that way, which I'm not sure we should.)

On consumer hardware probably no way to turn off processors, on enterprise level hardware like sparc64 maybe you can

dbwiddis commented:
Any update on whether my disk name and size non-root backup worked for you?

I think this is ready to merge.  If you do, click "Ready for review" above!  Or click squash and merge yourself! :)

mprins commented:
LGTM
```
Disks:
 sd0: (model: ATA, SAMSUNG MZ7TD256, DXT0 - S/N: unknown) size: 256.1 GB, reads: 191321 (6.6 GiB), writes: 267191 (3.9 GiB), xfer: 139200
 |-- /dev/sd0a: sd0a (unknown) Maj:Min=4:0, size: 1.1 GB @ /
 |-- /dev/sd0l: sd0l (unknown) Maj:Min=4:11, size: 196.7 GB @ /home
 |-- /dev/sd0d: sd0d (unknown) Maj:Min=4:3, size: 4.2 GB @ /tmp
 |-- /dev/sd0f: sd0f (unknown) Maj:Min=4:5, size: 2.1 GB @ /usr
 |-- /dev/sd0g: sd0g (unknown) Maj:Min=4:6, size: 1.1 GB @ /usr/X11R6
 |-- /dev/sd0h: sd0h (unknown) Maj:Min=4:7, size: 10.6 GB @ /usr/local
 |-- /dev/sd0k: sd0k (unknown) Maj:Min=4:10, size: 6.3 GB @ /usr/obj
 |-- /dev/sd0j: sd0j (unknown) Maj:Min=4:9, size: 2.1 GB @ /usr/src
 |-- /dev/sd0e: sd0e (unknown) Maj:Min=4:4, size: 20.9 GB @ /var
 cd0: (model: MATSHITA, DVD-RAM UJ8E2, SB01 - S/N: unknown) size: ?, reads: ? (?), writes: ? (?), xfer: ?
 sd1: (model: KingSpec, Z3-128, 1214 - S/N: unknown) size: 128.0 GB, reads: 33 (42.6 KiB), writes: 0 (0 bytes), xfer: 0
```
